cmd/bucket-lifecycle-handlers.go

```diff
@@ -86,6 +87,39 @@ func (api objectAPIHandlers) PutBucketLifecycleHandler(w http.ResponseWriter, r
 		return
 	}
 
+	// Create a map of updated set of rules in request
+	updatedRules := make(map[string]lifecycle.Rule, len(bucketLifecycle.Rules))
+	for _, rule := range bucketLifecycle.Rules {
+		updatedRules[rule.ID] = rule
+	}
+
+	// Get list of rules for the bucket from disk
+	meta, err := globalBucketMetadataSys.GetConfigFromDisk(ctx, bucket)
+	if err != nil {
+		writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL)
+		return
+	}
+	expiryRuleRemoved := false
+	if len(meta.LifecycleConfigXML) > 0 {
+		var lcCfg lifecycle.Lifecycle
+		if err := xml.Unmarshal(meta.LifecycleConfigXML, &lcCfg); err != nil {
+			writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL)
+			return
+		}
+		for _, rl := range lcCfg.Rules {
+			// if rule not found the incoming rules from request and was an expiry rule
+			// set the flag that expiry rule removed as part of request
+			if _, ok := updatedRules[rl.ID]; ok && !rl.Expiration.IsNull() || !rl.NoncurrentVersionExpiration.IsNull() {
+				expiryRuleRemoved = true
+			}
```

Comment:
this check is flawed - if a rule had just transition flags and is unchanged, expiryRuleRemoved would be true.

```
updtdRule, ok := updatedRules[rl.ID]
// original rule had expiry that is no longer in the new config, or rule is present but missing expiration flags
if (!rl.Expiration.IsNull() || !rl.NoncurrentVersionExpiration.IsNull()) && 
   (!ok || (updtdRuleExpiration.IsNull() && updtdRule.NoncurrentVersionExpiration.IsNull()){
				expiryRuleRemoved = true
}
```

Reply:
Ah yes. didn't realize that. Will update. Thanks!